src/main/java/ch/cern/eam/wshub/core/services/administration/impl/UserGroupMenuServiceImpl.java

```diff
@@ -0,0 +1,321 @@
+                if (!gmeFound && gme.description.equals(next) && (current == null || gme.getParent().equals(current.getId()))) {
+                    current = gme;
+                    gmeFound = true;
+                    amountFound++;
+                }
+            }
+            if (!gmeFound) {
+                return Arrays.copyOf(words, amountFound);
+            }
+        }
+
+        return words; // If reached the end, then all path could be found
+    }
+
+
+    /**
+     * Adds a full menu/submenu/function path to the menu hierarchy.
+     * @param context the user credentials
+     * @param node the specified full path and function to add, for a specific user group
+     * @return
+     */
+    @Override
+    public String addToMenuHierarchy(InforContext context, MenuSpecification node) throws InforException {
+        this.validateInputNode(node);
+
+        // Get menu entries as list
+        List<GenericMenuEntry> menuEntries = this.getExtMenuHierarchyAsList(context, node);
+
+        // Check if path already exists; if so, continue
+        String[] words = node.path.split("\\/");
+        String[] existingPath = this.calculateExistingPath(words, menuEntries);
+
+        if (words.length - existingPath.length == 0 && node.menuCode.isEmpty()) { // Path already exists and no function to add
+            return "OK";
+        }
+
+        // Check if path is incomplete; if so, complete it (if it doesn't, add all submenus (or menu)) starting from second to last item
+        if ((words.length - existingPath.length) > 0) { // Path is incomplete (or doesn't exist), since differs more than 0
+            UserGroupMenuService.MenuSpecification recursiveMenuSpecification = new UserGroupMenuService.MenuSpecification(node.path, "", node.userGroup); // If function is there, then we remove only that
+            if (node.menuCode.isEmpty()) { // But if no function, then we remove only one path item at the end
+                recursiveMenuSpecification = new UserGroupMenuService.MenuSpecification(String.join("/", Arrays.copyOf(words, words.length - 1)), node.menuCode, node.userGroup);
+            }
+            this.addToMenuHierarchy(context, recursiveMenuSpecification);
+            menuEntries = this.getExtMenuHierarchyAsList(context, recursiveMenuSpecification); //TODO Calling a get two times is not good; fixed by using a HashMap or a tree in a future iteration
+        }
+
+        // Now add leaf item
+        // Decide menu type: if function type is not set, then assume it's a menu and not a function
+        String menuType = this.decideMenuType(node.menuCode, words);
+
+        // Find parent id of new entry/item from previous list
+        GenericMenuEntry parent;
+        if (menuType.equals("M") || menuType.equals("F")) { //TODO better with an enum, even if it's not our design (but might be too cluttered)
+            parent = this.getEntryByPathFromList(menuEntries, Arrays.copyOf(words, words.length - 1));
```

Comment:
If instead of path we had a GenericMenuEntry here, we wouldn't need to handle these cases differently. In fact, I think the menu entry structure is clearly a tree structure and, if modeled as such, would make most of these operations simpler and less bug prone. The only time we might require a list is when creating a new entry, and even then receiving a tree structure with only the information required works

Reply:
Now far more clear with tree structure. getEntryByPathFromList no longer needed, deleted.